File: week_25/05_parallel_http.rb
#!/usr/bin/env ruby
require 'net/http'
require 'uri'
require 'json'

# 05. Parallel HTTP Requests in Ruby

puts "=== Ruby Parallel HTTP Requests ==="
puts

urls = [
  'https://jsonplaceholder.typicode.com/posts/1',
  'https://jsonplaceholder.typicode.com/posts/2',
  'https://jsonplaceholder.typicode.com/posts/3',
  'https://jsonplaceholder.typicode.com/posts/4',
  'https://jsonplaceholder.typicode.com/posts/5'
]

# Sequential (slow)
puts "1. Sequential requests:"
start_time = Time.now

urls.each do |url|
  response = Net::HTTP.get_response(URI(url))
  puts "  #{url}: #{response.code}"
end

sequential_time = Time.now - start_time
puts "  Time: #{sequential_time.round(2)}s"
puts

# Parallel (fast)
puts "2. Parallel requests (with threads):"
start_time = Time.now

threads = urls.map do |url|
  Thread.new(url) do |u|
    response = Net::HTTP.get_response(URI(u))
    puts "  #{u}: #{response.code}"
    response
  end
end

responses = threads.map(&:value)  # Wait and collect results
parallel_time = Time.now - start_time

puts "  Time: #{parallel_time.round(2)}s"
puts "  Speedup: #{(sequential_time / parallel_time).round(2)}x"
puts

# With error handling
puts "3. With error handling:"

def fetch_url(url)
  uri = URI(url)
  response = Net::HTTP.get_response(uri)
  
  if response.code == '200'
    data = JSON.parse(response.body)
    { success: true, title: data['title'], url: url }
  else
    { success: false, error: "HTTP #{response.code}", url: url }
  end
rescue => e
  { success: false, error: e.message, url: url }
end

threads = urls.map do |url|
  Thread.new { fetch_url(url) }
end

results = threads.map(&:value)

results.each do |result|
  if result[:success]
    puts "  ✓ #{result[:url]}"
    puts "    Title: #{result[:title][0..50]}..."
  else
    puts "  ✗ #{result[:url]}: #{result[:error]}"
  end
end

puts
puts "✅ Parallel HTTP with error handling complete"

# Key points:
# - Threads good for I/O (HTTP, DB)
# - thread.value waits and returns result
# - Much faster than sequential
# - Handle errors per thread
# - GIL doesn't matter for I/O

File: week_25/05_parallel_http.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sync"
	"time"
)

// 05. Parallel HTTP Requests in Go

type Post struct {
	ID    int    `json:"id"`
	Title string `json:"title"`
}

func main() {
	fmt.Println("=== Go Parallel HTTP Requests ===")
	fmt.Println()

	urls := []string{
		"https://jsonplaceholder.typicode.com/posts/1",
		"https://jsonplaceholder.typicode.com/posts/2",
		"https://jsonplaceholder.typicode.com/posts/3",
		"https://jsonplaceholder.typicode.com/posts/4",
		"https://jsonplaceholder.typicode.com/posts/5",
	}

	// Sequential (slow)
	fmt.Println("1. Sequential requests:")
	start := time.Now()

	for _, url := range urls {
		resp, _ := http.Get(url)
		fmt.Printf("  %s: %s\n", url, resp.Status)
		resp.Body.Close()
	}

	sequentialTime := time.Since(start)
	fmt.Printf("  Time: %.2fs\n", sequentialTime.Seconds())
	fmt.Println()

	// Parallel (fast)
	fmt.Println("2. Parallel requests (with goroutines):")
	start = time.Now()

	var wg sync.WaitGroup
	for _, url := range urls {
		wg.Add(1)
		go func(u string) {
			defer wg.Done()
			resp, _ := http.Get(u)
			fmt.Printf("  %s: %s\n", u, resp.Status)
			resp.Body.Close()
		}(url)
	}

	wg.Wait()
	parallelTime := time.Since(start)

	fmt.Printf("  Time: %.2fs\n", parallelTime.Seconds())
	fmt.Printf("  Speedup: %.2fx\n", sequentialTime.Seconds()/parallelTime.Seconds())
	fmt.Println()

	// With error handling
	fmt.Println("3. With error handling:")

	type Result struct {
		Success bool
		Title   string
		URL     string
		Error   string
	}

	resultCh := make(chan Result, len(urls))

	for _, url := range urls {
		wg.Add(1)
		go func(u string) {
			defer wg.Done()

			resp, err := http.Get(u)
			if err != nil {
				resultCh <- Result{Success: false, Error: err.Error(), URL: u}
				return
			}
			defer resp.Body.Close()

			if resp.StatusCode != 200 {
				resultCh <- Result{Success: false, Error: resp.Status, URL: u}
				return
			}

			body, _ := io.ReadAll(resp.Body)
			var post Post
			if err := json.Unmarshal(body, &post); err != nil {
				resultCh <- Result{Success: false, Error: err.Error(), URL: u}
				return
			}

			resultCh <- Result{Success: true, Title: post.Title, URL: u}
		}(url)
	}

	// Close result channel after all goroutines done
	go func() {
		wg.Wait()
		close(resultCh)
	}()

	// Collect results
	for result := range resultCh {
		if result.Success {
			fmt.Printf("  ✓ %s\n", result.URL)
			title := result.Title
			if len(title) > 50 {
				title = title[:50] + "..."
			}
			fmt.Printf("    Title: %s\n", title)
		} else {
			fmt.Printf("  ✗ %s: %s\n", result.URL, result.Error)
		}
	}

	fmt.Println()
	fmt.Println("✅ Parallel HTTP with error handling complete")
}

// Key points:
// - Goroutines are lightweight
// - Channels to collect results
// - close(ch) after wg.Wait()
// - Struct for result with error
// - Very efficient for I/O
// - No GIL - true parallelism
